src/casp/services/model_inference/services.py

```diff
@@ -18,34 +19,81 @@ def __init__(self):
-        return pickle.loads(pickle_in)
+        return f"gs://{settings.PROJECT_BUCKET_NAME}/{model_checkpoint_file_path}"
+
+    def _get_model_checkpoint_file(self, model: models.CASModel) -> t.BinaryIO:
+        """
+        Get model checkpoint from either local or GCS and load it using CellariumModule.
+
+        :param model: Cellarium Cloud model db object
+
+        :return: CellariumModule object
+        """
+        model_checkpoint_path = self._get_model_checkpoint_path(model.model_file_path)
+
+        return open(model_checkpoint_path, "rb")
+
+    def _get_output_from_model(
+        self, model: models.CASModel, adata_file: t.BinaryIO
+    ) -> t.Tuple[numpy.ndarray, t.List[str]]:
+        """
+        Get output from cellarium-ml model that predicts embeddings given an input adata file.
+
+        :param model: Cellarium Cloud model db object
+        :param adata_file: File object of :class:`anndata.AnnData` object to embed.
+
+        :return: Tuple of embeddings and obs_ids.
+        """
+        adata = anndata.read_h5ad(adata_file)
+
+        cellarium_checkpoint_file = self._get_model_checkpoint_file(model)
+        cellarium_module = CellariumModule.load_from_checkpoint(cellarium_checkpoint_file, map_location="cpu")
+        cellarium_checkpoint_file.seek(0)
+
+        cellarium_data_module = CellariumAnnDataDataModule.load_from_checkpoint(
+            cellarium_checkpoint_file, dadc=adata, batch_size=adata.n_obs, num_workers=0
+        )
+        cellarium_data_module.setup()
+        batch = next(iter(cellarium_data_module.train_dataloader()))
+
+        cellarium_output_dict = cellarium_module(batch)
+
+        embeddings = cellarium_output_dict["x_ng"].numpy()
+
+        obs_ids = adata.obs.index.tolist()
+        return embeddings, obs_ids
 
     @staticmethod
-    def load_data(file: t.BinaryIO) -> t.Tuple[torch.Tensor, list]:
+    def _validate_model_output(embeddings: np.ndarray, obs_ids: t.List[str], model_info: models.CASModel) -> None:
         """
-        Load Sparse Matrix and ids from anndata file
+        Validate model output.
 
-        :param file: File of :class:`anndata.Anndata` object to load
+        :param embeddings: Embeddings
+        :param obs_ids: List of observation ids
+        :param model_info: Cellarium Cloud model db object
 
-        :return: Tuple of Sparse Matrix and ids
+        :raises ModelOutputError: If the length of obs_ids and embeddings are not the same or if the number of embedding
+            dimensions is not equal to the model's embedding dimensions.
         """
-        adata = anndata.read_h5ad(file)
-        X = torch.Tensor(adata.X.todense().astype(int))
-        obs_ids = adata.obs.index.values.astype(str).tolist()
-        return X, obs_ids
+        if embeddings.shape[0] != len(obs_ids):
+            raise exceptions.ModelOutputError(
+                "Model Output Error: Length of obs_ids and embeddings should be the same."
```

Comment:
Would it be valuable to include the lengths for `obs_ids` and `embeddings` in this message?

Reply:
Absolutely, yes! I will add it.